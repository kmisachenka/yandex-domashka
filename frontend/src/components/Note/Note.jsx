import React from 'react';
import * as PropTypes from 'prop-types';
import hexToRgba from 'hex-to-rgba';
import cn from 'classnames';

import styles from './Note.module.scss';

import Reminder from '../Reminder';
import Attachments from '../Attachements';
import ListSection from './ListSection';
import TextSection from './TextSection';
import ImageSection from './ImageSection';

export default function Note(props) {
  const { note, colors } = props;
  const { reminder, attachments, size } = note;

  const getColorOrDefault = id => {
    if (id !== undefined) {
      const { color } = colors.find(o => o.id === id);
      const rgba = hexToRgba(color, 0.4);
      return `linear-gradient(0deg, ${rgba}, ${rgba}), #FFFFFF`;
    }
    return '#FFFFFF';
  };

  const renderByType = noteToBeRendered => {
    const { type, color } = noteToBeRendered;
    const background = getColorOrDefault(color);
    switch (type) {
      case 'text':
        return <TextSection note={noteToBeRendered} background={background} />;
      case 'list':
        return <ListSection note={noteToBeRendered} background={background} />;
      case 'image':
        return <ImageSection note={noteToBeRendered} background={background} />;
      default:
        return <></>;
    }
  };

  return (
    <article className={cn(styles.note, styles[size])}>
      {reminder && (
        <section className={styles.reminder}>
          <Reminder reminder={reminder} />
        </section>
      )}
      {renderByType(note)}
      {attachments && attachments.length && (
        <section className={styles.attachments}>
          <Attachments attachments={attachments} />
        </section>
      )}
    </article>
  );
}

Note.propTypes = {
  note: PropTypes.shape({
    type: PropTypes.oneOf(['list', 'text', 'image']).isRequired,
    size: PropTypes.oneOf(['s', 'm', 'l']).isRequired,
    created: PropTypes.number.isRequired,
    title: PropTypes.string,
    text: PropTypes.string,
    tags: PropTypes.array,
    color: PropTypes.number,
    attachments: PropTypes.arrayOf(
      PropTypes.shape({
        type: PropTypes.oneOf(['link', 'image']).isRequired,
        url: PropTypes.string.isRequired,
      })
    ),
  }).isRequired,
  colors: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      color: PropTypes.string.isRequired,
    })
  ).isRequired,
};
