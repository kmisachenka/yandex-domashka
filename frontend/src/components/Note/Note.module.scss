@import '../../media';
@import '../NoteFooter/NoteFooter.module';

.s {
  @include tablet {
    grid-column: span 2;
    grid-row: span 1;
  }

  @include desktop {
    grid-column: span 2;
    grid-row: span 1;

    &:last-child {
      grid-row: span 2;
    }
  }
}

.m {
  grid-row: span 2;

  @include tablet {
    grid-column: span 2;
  }

  @include desktop {
    grid-column: span 3;
  }
}

.l {
  grid-row: span 2;

  @include tablet {
    grid-column: span 2;
  }

  @include desktop {
    grid-column: span 4;
    grid-row: span 3;
  }
}

.note {
  border: 1px solid rgba(218, 220, 224, 0.5);
  border-radius: 16px;
  display: flex;
  flex-direction: column;

  height: 100%;
  justify-content: space-between;

  position: relative;

  z-index: 10;
}

.body {
  padding-left: 24px;
  padding-right: 24px;
  padding-top: 24px;
}

.reminder {
  align-items: center;
  background: #db5341;
  border: 1px solid rgba(218, 220, 224, 0.5);
  border-radius: 16px;

  display: flex;
  margin-bottom: -24px;

  padding: 16px 24px 36px;
}

.attachments {
  align-content: center;
  align-items: center;
  background: rgba(216, 216, 216, 0.5);
  border-radius: 16px;

  display: flex;

  margin-top: -24px;

  padding: 36px 12px 12px;
}
