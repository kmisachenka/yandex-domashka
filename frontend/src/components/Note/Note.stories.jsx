import React from 'react';
import { storiesOf } from '@storybook/react';
import Note from './Note';
import Provider from '../reduxProvider';

const colors = [
  {
    id: 0,
    color: '#E84747',
  },
  {
    id: 1,
    color: '#F2994A',
  },
  {
    id: 2,
    color: '#F2C94C',
  },
  {
    id: 3,
    color: '#219653',
  },
  {
    id: 4,
    color: '#56CCF2',
  },
  {
    id: 5,
    color: '#2F80ED',
  },
  {
    id: 6,
    color: '#9B51E0',
  },
];

storiesOf('Заметка', module)
  .addDecorator(story => <Provider story={story()} />)
  .add('Завтра дедлайн', () => {
    const note = {
      type: 'text',
      title: 'Завтра дедлайн',
      text: 'Не забыть сверстать макет',
      tags: [1, 2],
      color: 2,
      size: 's',
      created: 1551714600000,
    };

    return <Note colors={colors} note={note} />;
  })
  .add('Купить чайник', () => {
    const note = {
      type: 'text',
      text: 'Купить чайник с Bluetooth\nВидел у Xiaomi на днях',
      tags: [0],
      size: 's',
      created: 1551714600000,
    };

    return <Note colors={colors} note={note} />;
  })
  .add('Сиба-Ин', () => {
    const note = {
      type: 'text',
      title: 'Не забыть выгулять Сиба-Ину',
      color: 3,
      size: 's',
      created: 1520160803000,
    };

    return <Note colors={colors} note={note} />;
  })
  .add('Выучить латынь', () => {
    const note = {
      type: 'text',
      title: 'Выучить латынь',
      text:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
      tags: [7, 8],
      size: 'l',
      created: 1543798200000,
    };

    return <Note colors={colors} note={note} />;
  });
