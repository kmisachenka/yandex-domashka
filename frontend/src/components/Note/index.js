import Note from './Note';

export { Note as default };
