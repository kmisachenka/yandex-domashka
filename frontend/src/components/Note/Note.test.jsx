import React from 'react';
import { shallow } from 'enzyme';

import Note from './Note';

const note = {
  id: 3,
  type: 'text',
  text: 'Купить чайник с Bluetooth\nВидел у Xiaomi на днях',
  tags: [0],
  size: 's',
  created: 1551593220000,
};

describe('Note', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<Note note={note} color="#FFFFFF" />);
    expect(wrapper).toMatchSnapshot();
  });
});
