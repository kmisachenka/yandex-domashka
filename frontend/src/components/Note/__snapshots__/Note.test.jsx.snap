// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note should match snapshot 1`] = `
<article
  className="note s"
>
  <Text />
</article>
`;
